src/NServiceBus.Transport.Msmq/MessagePump.cs

```diff
@@ -192,7 +216,7 @@ async Task PumpMessages(CancellationToken messagePumpCancellationToken)
         }
 
         // This is static to prevent the method from accessing fields in the pump since that causes variable capturing and cause extra allocations
-        static async Task ReceiveMessagesSwallowExceptionsAndReleaseConcurrencyLimiter(MessagePump messagePump, CancellationToken messagePumpCancellationToken)
+        static async Task ReceiveMessagesSwallowExceptionsAndReleaseConcurrencyLimiter(MessagePump messagePump, SemaphoreSlim localConcurrencyLimiter, CancellationToken messagePumpCancellationToken)
```

Comment:
`localConcurrencyLimiter` maybe is irrelevant. How about `concurrencyLimiter`. Also why is this change needed? Following the logic you could also assign extract it from the pump at the beginning, assign it to a local and release that one at the end couldn't you?

Reply:
I think it is needed. We need to release the same semaphore as we acquired in the previous method to make sure we drain all semaphores correctly. So I'll fix the passing of the semaphore to use the tuple.